mono/mini/ssa.c

```diff
@@ -826,6 +826,10 @@ evaluate_ins (MonoCompile *cfg, MonoInst *ins, MonoInst **res, MonoInst **carray
 		return 2;
 
 	num_sregs = mono_inst_get_src_registers (ins, sregs);
+
+	if (num_sregs > 2)
+		return 2;
```

Comment:
I think this is a bit too hack-ish. There is no real reason to _not_ support constant folding instructions with `num_sregs = 3`. The reason why we assert on it currently is that we do not have any instruction with that property where we could constant fold.

Why do we hit the assert then? First of all let us figure out why this only hits on 32bit. If you look at the generated IR for your testcase, you get on amd64:
```
 il_seq_point il: 0x7
 i8const R21 <- [4361201612]
 iconst R22 <- [1]
 iconst R23 <- [0]
 atomic_cas_i4 R25 <- [R21 + 0x0] R22 R23
```
vs. x86:
```
 iconst R13 <- [11144108]
 iconst R14 <- [1]
 iconst R15 <- [0]
 atomic_cas_i4 R17 <- [R13 + 0x0] R14 R15
```

The assert triggers for `atomic_cas_i4` only on 32bit because `i8const` isn't properly detected as constant. In fact, we need to extend this https://github.com/mono/mono/blob/a76fc055cf1dc3173b6df1690d14723a13fbaf63/mono/mini/ssa.c#L819-L823
to also support `OP_I8CONST`
```c
      /* Short-circuit this */
      if (ins->opcode == OP_ICONST || ins->opcode == OP_I8CONST) {
          *res = ins;
          return 1;
      }
```
_then_ you get the same issue on 64bit. <sup>1</sup>

So the assert hits when processing `atomic_cas_i4 R25 <- [R21 + 0x0] R22 R23`. I think the loop below should be able to detect memory access on a constant address and set `const_args = FALSE` in such cases. If we ever add instructions with `num_sregs = 3` and a constant folding opportunity would appear, we would still hit the `g_assert (num_sregs <= 2);`, letting us know to do something about it. With your change we would silently skip over it.


<sup>1</sup> Unfortunately this seems to break other things _sigh_

Reply:
Yikes. If any of this remains unfixed, this should be documented with comments in the source. Sounds like something that would produce difficult-to-discern causes if it crashed in the field. 

@lewurm what other things does it break? Maybe a separate PR could consider those.

Reply:
It is true that there's no reason not to support constant folding ternary instructions.

Some ternary operations could benefit from constant folding--the ones that come to mind are fused multiply-add and SIMD blends/shuffles. But I don't think a crashing assertion is appropriate for this, even if we add constant-foldable ternary instructions, because I'm not sure what "abstract" invariant is being violated here. For example, this code doesn't crash when it encounters 64-bit arithmetic operations (LMUL/LADD/etc.); it just silently ignores them.

One less hacky option would be to extend `mono_constant_fold_ins` to take a pointer to an array of `MonoInst *[MONO_MAX_SRC_REGS]`; `mono_constant_fold_ins` could then, for the time being, treat ternary operations the same way it treats 64-bit arithmetic operations (i.e., it could ignore them). The failing assertion could then be completely removed.

Reply:
We don't have ternary ops which could be cfold-ed right now, so this looks fine to me.

Reply:
> But I don't think a crashing assertion is appropriate for this, even if we add constant-foldable ternary instructions, because I'm not sure what "abstract" invariant is being violated here.

True, an assertion might be too... assertive 🙂  There's `g_warning ()`, but depending on the situation that might be too annoying in the long-term.

For the sake of fixing the bug we can go with the suggested change, but please add a comment to it.

We should follow up on the `I8CONST` thing though, looks like we miss some constant folding opportunities.